sys/include/net/sock/ip.h

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (C) 2016 Alexander Aring <aar@pengutronix.de>
+ *                    Freie Universität Berlin
+ *                    HAW Hamburg
+ *                    Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_sock_ip     Raw IPv4/IPv6 sock API
+ * @ingroup     net_sock
+ * @brief       Sock Submodule for raw IPv4/IPv6
+ * @{
+ *
+ * @file
+ * @brief   Raw IPv4/IPv6 sock definitions
+ *
+ * @author  Alexander Aring <aar@pengutronix.de>
+ * @author  Simon Brummer <simon.brummer@haw-hamburg.de>
+ * @author  Cenk Gündoğan <mail@cgundogan.de>
+ * @author  Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ */
+#ifndef NET_SOCK_IP_H_
+#define NET_SOCK_IP_H_
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "net/sock/addr.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   An end point for a raw IPv4/IPv6 sock object
+ */
+typedef struct {
+    sock_addr_ip_t addr;    /**< IP address */
+    int family;             /**< family of sock_ip_ep_t::addr */
+
+    /**
+     * @brief   network interface ID as defined in @ref net_netif
+     *
+     * Use @ref SOCK_ADDR_ANY_NETIF for any interface.
+     * For reception this is the local interface the message came over,
+     * for transmission, this is the local interface the message should be send
+     * over
+     */
+    uint16_t netif;
+} sock_ip_ep_t;
+
+/**
+ * @brief   Implementation-specific type of a raw IPv4/IPv6 sock object
+ *
+ * `struct sock_ip` needs to be defined by stack-specific a implementation.
+ */
+typedef struct sock_ip sock_ip_t;
+
+/**
+ * @brief   Creates a new raw IPv4/IPv6 sock object
+ *
+ * @pre `(sock != NULL)`
+ *
+ * @param[out] sock     Preallocated sock object. Must fill the size of
+ *                      the stack-specific sock descriptor.
```

Comment:
No. The function fills it. But if the type is not of the size the function expects (which might well be if you put the wrong `sock_ip_t` in here) this can very well lead to unexpected behavior.